File: plugin/hooks/src/io/github/cowwoc/cat/hooks/GetBashPretoolOutput.java
package io.github.cowwoc.cat.hooks;

import static io.github.cowwoc.cat.hooks.Strings.equalsIgnoreCase;
import static io.github.cowwoc.requirements13.java.DefaultJavaValidators.requireThat;

import io.github.cowwoc.cat.hooks.bash.BlockLockManipulation;
import io.github.cowwoc.cat.hooks.bash.BlockMainRebase;
import io.github.cowwoc.cat.hooks.bash.BlockMergeCommits;
import io.github.cowwoc.cat.hooks.bash.BlockReflogDestruction;
import io.github.cowwoc.cat.hooks.bash.ComputeBoxLines;
import io.github.cowwoc.cat.hooks.bash.RemindGitSquash;
import io.github.cowwoc.cat.hooks.bash.ValidateCommitType;
import io.github.cowwoc.cat.hooks.bash.ValidateGitFilterBranch;
import io.github.cowwoc.cat.hooks.bash.ValidateGitOperations;
import io.github.cowwoc.cat.hooks.bash.WarnFileExtraction;
import tools.jackson.databind.JsonNode;

import java.util.ArrayList;
import java.util.List;

/**
 * get-bash-pretool-output - Unified PreToolUse hook for Bash commands.
 *
 * <p>TRIGGER: PreToolUse (matcher: Bash)</p>
 *
 * <p>Consolidates all Bash command validation hooks into a single Java dispatcher.</p>
 *
 * <p>Handlers can:</p>
 * <ul>
 *   <li>Block commands (return decision=block with reason)</li>
 *   <li>Warn about commands (return warning)</li>
 *   <li>Allow commands (return allow)</li>
 * </ul>
 */
public final class GetBashPretoolOutput
{
  private GetBashPretoolOutput()
  {
    // Utility class
  }

  private static final List<BashHandler> HANDLERS = List.of(
    new BlockLockManipulation(),
    new BlockMainRebase(),
    new BlockMergeCommits(),
    new BlockReflogDestruction(),
    new ComputeBoxLines(),
    new RemindGitSquash(),
    new ValidateCommitType(),
    new ValidateGitFilterBranch(),
    new ValidateGitOperations(),
    new WarnFileExtraction());

  /**
   * Entry point for the Bash pretool output hook.
   *
   * @param _args command line arguments (unused)
   */
  @SuppressWarnings("UnusedVariable")
  public static void main(String[] _args)
  {
    HookInput input = HookInput.readFromStdin();

    String toolName = input.getToolName();
    if (!equalsIgnoreCase(toolName, "Bash"))
    {
      HookOutput.empty();
      return;
    }

    JsonNode toolInput = input.getToolInput();
    JsonNode commandNode = toolInput.get("command");
    String command;
    if (commandNode != null)
      command = commandNode.asString();
    else
      command = "";

    if (command.isEmpty())
    {
      HookOutput.empty();
      return;
    }

    String sessionId = input.getSessionId();
    requireThat(sessionId, "sessionId").isNotBlank();
    List<String> warnings = new ArrayList<>();

    // Run all bash pretool handlers
    for (BashHandler handler : HANDLERS)
    {
      try
      {
        BashHandler.Result result = handler.check(command, toolInput, null, sessionId);
        if (result.blocked())
        {
          // Handler blocked the command
          HookOutput.block(result.reason(), result.additionalContext());
          return;
        }
        if (!result.reason().isEmpty())
          warnings.add(result.reason());
      }
      catch (Exception e)
      {
        System.err.println("get-bash-pretool-output: handler error: " + e.getMessage());
      }
    }

    // Output warnings if any
    for (String warning : warnings)
    {
      System.err.println(warning);
    }

    // Allow the command
    HookOutput.empty();
  }
}

File: plugin/hooks/get-bash-pretool-output.py
#!/usr/bin/env python3
"""
get-bash-pretool-output.py - Unified PreToolUse hook for Bash commands

TRIGGER: PreToolUse (matcher: Bash)

Consolidates all Bash command validation hooks into a single Python dispatcher.

Handlers can:
- Block commands (return {"decision": "block", "reason": "..."})
- Warn about commands (return {"warning": "..."})
- Allow commands (return None)

Benefits:
1. Single Python startup vs N shell startups
2. Native JSON parsing
3. Shared pattern matching
4. Better error handling
"""

import json
import sys
from pathlib import Path

# Add handlers directory to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from bash_handlers import get_all_handlers


def read_hook_input() -> dict:
    """Read and parse JSON input from stdin."""
    if sys.stdin.isatty():
        return {}

    try:
        raw = sys.stdin.read()
        return json.loads(raw) if raw.strip() else {}
    except json.JSONDecodeError:
        return {}


def output_block(reason: str, additional_context: str = None) -> None:
    """Output a block decision."""
    response = {
        "decision": "block",
        "reason": reason
    }
    if additional_context:
        response["additionalContext"] = additional_context
    print(json.dumps(response))


def output_warning(warning: str) -> None:
    """Output a warning (stderr) but allow the command."""
    sys.stderr.write(warning + "\n")
    print("{}")


def main():
    # Read hook input
    hook_data = read_hook_input()

    # Extract tool info
    tool_name = hook_data.get("tool_name", "")
    if tool_name.lower() != "bash":
        print("{}")
        return

    tool_input = hook_data.get("tool_input", {})
    command = tool_input.get("command", "")

    if not command:
        print("{}")
        return

    # Build context for handlers
    # CRITICAL (M360): Include cwd from hook_data so handlers can check shell's working directory
    context = {
        "tool_input": tool_input,
        "session_id": hook_data.get("session_id", ""),
        "cwd": hook_data.get("cwd", ""),  # M360: Pass cwd for worktree validation
        "hook_data": hook_data,
    }

    # Run all handlers
    warnings = []
    for handler in get_all_handlers():
        try:
            result = handler.check(command, context)
            if result:
                # Block decision takes priority
                if result.get("decision") == "block":
                    output_block(
                        result.get("reason", "Command blocked"),
                        result.get("additionalContext")
                    )
                    return
                # Collect warnings
                if result.get("warning"):
                    warnings.append(result["warning"])
        except Exception as e:
            sys.stderr.write(f"get-bash-pretool-output: handler error: {e}\n")

    # Output warnings if any
    if warnings:
        for warning in warnings:
            sys.stderr.write(warning + "\n")

    # Allow the command
    print("{}")


if __name__ == "__main__":
    main()
